public/src/main/java/com/rackspace/salus/telemetry/api/services/TestMonitorAndEventTaskService.java

```diff
@@ -39,56 +41,82 @@
-            List<SimpleNameTagValueMetric> metrics = testMonitorResult.getData().getMetrics().stream()
-                    .filter(
-                            e -> e.getName().equals(testMonitorAndEventTaskRequest.getTask().getMeasurement()))
-                    .collect(Collectors.toList());
-            if (CollectionUtils.isEmpty(metrics)) {
-                return new TestMonitorAndEventTaskResponse(
-                        new TestMonitorAndEventTask(testMonitorResult, null),
-                        List.of("Unable to find matching metric name"));
-            }
-            TestTaskRequest testTaskRequest = new TestTaskRequest()
-                    .setTask(testMonitorAndEventTaskRequest.getTask())
-                    .setMetrics(metrics);
-            testTaskResult = eventTaskApi
-                    .performTestTask(tenantId, testTaskRequest);
-            if (!CollectionUtils.isEmpty(testTaskResult.getErrors()) || testTaskResult.getData() == null) {
-                return new TestMonitorAndEventTaskResponse(
-                        new TestMonitorAndEventTask(testMonitorResult, testTaskResult),
-                        CollectionUtils.isEmpty(testTaskResult.getErrors()) ? List.of("Unable to test-events") : testTaskResult.getErrors());
-            }
+      testTaskResult = eventTaskApi
+          .performTestTask(tenantId, testTaskRequest);
 
-            return new TestMonitorAndEventTaskResponse(
-                    new TestMonitorAndEventTask(testMonitorResult, testTaskResult), null);
-        } catch (RemoteServiceCallException e) {
-            return new TestMonitorAndEventTaskResponse(
-                    new TestMonitorAndEventTask(testMonitorResult, testTaskResult), List.of(String
-                    .format("An unexpected internal error occurred: %s", e.getMessage())));
-        }
+      if (testTaskResult == null || !CollectionUtils.isEmpty(testTaskResult.getErrors())
+          || testTaskResult.getData() == null) {
+        return new TestMonitorAndEventTaskResponse().setData(new TestMonitorAndEventTask()
+            .setMonitor(
+                new TestMonitorResultData().setMetrics(testMonitorResult.getData().getMetrics())))
+            .setErrors(CollectionUtils.isEmpty(testMonitorResult.getErrors()) ? List
+                .of("Unable to get test event data") : testMonitorResult.getErrors());
+      }
 
+      return new TestMonitorAndEventTaskResponse().setData(new TestMonitorAndEventTask()
+          .setMonitor(
+              new TestMonitorResultData().setMetrics(testMonitorResult.getData().getMetrics()))
+          .setTask(new TestTaskResultResultData().setEvents(testTaskResult.getData().getEvents())
+              .setStats(testTaskResult.getData()
+                  .getStats())));
+    } catch (RemoteServiceCallException e) {
+      TestMonitorAndEventTaskResponse testMonitorAndEventTaskResponse = new TestMonitorAndEventTaskResponse();
+      if (testMonitorResult.getData() != null) {
```

Comment:
It seems like `testMonitorResult` might remain null if an exception occurred during that call.
```suggestion
      if (testMonitorResult != null && testMonitorResult.getData() != null) {
```